bls/src/main/java/tech/pegasys/teku/bls/impl/blst/BlstSignature.java

```diff
@@ -120,6 +120,13 @@ public Bytes toBytesCompressed() {
   @Override
   public boolean verify(List<PublicKeyMessagePair> keysToMessages) {
 
+    // according to BLS spec aggregateVerify allows only distinct messages
+    // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-3.1.1
+    if (keysToMessages.stream().map(PublicKeyMessagePair::getMessage).distinct().count()
+        < keysToMessages.size()) {
+      return false;
+    }
+
     boolean isAnyPublicKeyInfinity =
         keysToMessages.stream()
             .anyMatch(pair -> ((BlstPublicKey) pair.getPublicKey()).isInfinity());
```

Comment:
We're now iterating the list of messages twice, seems like it would be better to unroll these streams into an old style for loop and do both checks in a single pass.  That would also let us exit early if duplicate messages were found. ie

```
final Set<Bytes> seenMessages = new HashSet<>();
    for (PublicKeyMessagePair keysToMessage : keysToMessages) {
      // according to BLS spec aggregateVerify allows only distinct messages
      // https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-3.1.1
      if (!seenMessages.add(keysToMessage.getMessage())) {
        return false;
      }
      
      if (((BlstPublicKey)keysToMessage.getPublicKey()).isInfinity()) {
        return false;
      }
    }
```

Reply:
In the original conception of the BLS libs, only the methods in BLS.java are guaranteed to be standards-conforming; anything lower-level can be as hacky as we like.

Thus, probably the right thing to do is to leave the duplicate message check in BLS.java (where it [already exists](https://github.com/ConsenSys/teku/blob/master/bls/src/main/java/tech/pegasys/teku/bls/BLS.java#L149)), and omit it from the implementation.

Batch verification can then skip the duplicate message check, which will be perfectly ok since we should not be vulnerable to a rogue key attack. Also, batch verification is not in the standard, so we can implement it as we wish. When we run external tests we should hook them into the standards-conforming BLS.java methods, nothing lower level. Thoughts?

Reply:
My 2 cent. The test seem to target the [Basic scheme](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-3.1.1) while in Ethereum we are in the Pop realm (where duplicates are not an issue).
Now Teku doesn't do a clear separation between the 3 schemes specified in the draft (like e.g. `py_ecc` that implements all 3). I guess before to think about a fix (either delete the test or fix the code) you should decide what you what to support. cc @protolambda

Reply:
Thanks, @asanso Having looked more closely at the POP scheme, `aggregateVerify` under that [should be identical](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-3.3) to `coreAggregateVerify`, which I think I missed previously. This means that we should remove the duplicate message checks everywhere.

We're not aiming to provide a complete BLS implementation, only what's required for Teku/Eth2, so following the POP spec is fine and resolves all of the above discussion :-)

Reply:
Thanks @asanso, good catch 👍 
Removed 2 duplicate checks and corresponding unit tests